vpr/src/route/rr_graph_generation/rr_graph_sg.cpp

```diff
@@ -0,0 +1,352 @@
+
+    t_physical_tile_loc sb_loc0, sb_loc1;
+    switch (side) {
+        case TOP:
+            sb_loc0 = {x, y, layer};
+            sb_loc1 = {x - 1, y, layer};
+            break;
+
+        case BOTTOM:
+            sb_loc0 = {x, y - 1, layer};
+            sb_loc1 = {x - 1, y - 1, layer};
+            break;
+
+        case RIGHT:
+            sb_loc0 = {x, y, layer};
+            sb_loc1 = {x, y - 1, layer};
+            break;
+
+        case LEFT:
+            sb_loc0 = {x - 1, y, layer};
+            sb_loc1 = {x - 1, y - 1, layer};
+            break;
+
+        default:
+            VTR_ASSERT_SAFE(false);
+    }
+
+    const int grid_width = grid.width();
+    const int grid_height = grid.height();
+
+    sb_loc0.x = std::clamp(sb_loc0.x, 0, grid_width - 1);
+    sb_loc0.y = std::clamp(sb_loc0.y, 0, grid_height - 1);
+
+    sb_loc1.x = std::clamp(sb_loc1.x, 0, grid_width - 1);
+    sb_loc1.y = std::clamp(sb_loc1.y, 0, grid_height - 1);
+
+    std::vector<std::pair<RRNodeId, short>> selected_chanz_nodes0;
+    std::vector<std::pair<RRNodeId, short>> selected_chanz_nodes1;
+
+    for (int iseg = 0; iseg < num_seg_types; iseg++) {
+        int seg_index = get_parallel_seg_index(iseg, seg_index_map, e_parallel_axis::Z_AXIS);
+        if (seg_index < 0) {
+            continue;
+        }
+
+        selected_chanz_nodes0.clear();
+        for (size_t track_num = 0; track_num < interdie_3d_links[sb_loc0.x][sb_loc0.y].size(); track_num++) {
+            const t_bottleneck_link& bottleneck_link = interdie_3d_links[sb_loc0.x][sb_loc0.y][track_num];
+            if (bottleneck_link.parallel_segment_index == seg_index && bottleneck_link.gather_loc.layer_num == layer) {
+                RRNodeId node_id = node_lookup.find_node(sb_loc0.layer_num, sb_loc0.x, sb_loc0.y, e_rr_type::CHANZ, track_num);
+                selected_chanz_nodes0.emplace_back(node_id, bottleneck_link.arch_wire_switch);
+            }
+        }
+
+        selected_chanz_nodes1.clear();
+        for (size_t track_num = 0; track_num < interdie_3d_links[sb_loc1.x][sb_loc1.y].size(); track_num++) {
+            const t_bottleneck_link& bottleneck_link = interdie_3d_links[sb_loc1.x][sb_loc1.y][track_num];
+            if (bottleneck_link.parallel_segment_index == seg_index && bottleneck_link.gather_loc.layer_num == layer) {
+                RRNodeId node_id = node_lookup.find_node(sb_loc1.layer_num, sb_loc1.x, sb_loc1.y, e_rr_type::CHANZ, track_num);
+                selected_chanz_nodes1.emplace_back(node_id, bottleneck_link.arch_wire_switch);
+            }
+        }
+
+        for (RRNodeId opin_node_id : opin_nodes) {
+            int pin_number = rr_graph.node_pin_num(opin_node_id);
+            int fc = Fc_out[type->index][pin_number][iseg];
+
+            for (int i = 0; i < fc; i++) {
+
+                RRNodeId chanz_node_id;
+                short switch_id;
+                if (Fc_zofs[sb_loc0.x][sb_loc0.y][seg_index] < Fc_zofs[sb_loc1.x][sb_loc1.y][seg_index]) {
```

Comment:
Explain the logic here. I don't really understand the higher purpose of this selection of which node to use.